### ISSUE TITLE
SC2154 should optionally strict check to avoid errors when using "set -u"

### ISSUE THREAD

**Issue by kkmuffme** [koalaman/shellcheck#3081]
#### For bugs
- Rule Id 2154:
- My shellcheck version: 0.10.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
set -u
if [[ -n "$foo" ]]
then
	echo "$foo"
fi
```

#### Here's what shellcheck currently says:
No error

#### Here's what I wanted or expected to see:
>foo is referenced but not assigned
with code SC2154

Since this is an error.

Correct/no error code e.g.

```sh
#!/bin/bash
set -u
if [[ -n "${foo+x}" ]] && [[ -n "$foo" ]]
then
	echo "$foo"
fi
```

While usually this is pointless, in some cases the "set -u" is outside of your control (= project requirements, set by a parent file that calls your script,...) and of course these issues always happen when you're on vacation :-)
It would be helpful if there were a strict/pedantic mode for SC2154/check-unassigned-uppercase to check them as if "set -u" were set and report errors accordingly

**Comment by wileyhy** [comment]
`set -u` is such a test, and it's specified by POSIX. Why risk getting it
wrong while reimplementing it when the availability of nounset is already
required?

  Wiley

~ $ head -v test.sh
==> test.sh <==
#!/bin/bash -x
set -u
if [[ -n ${foo} ]]
then
        echo "$foo"
fi
~ $ ./test.sh
+ set -u
./test.sh: line 3: foo: unbound variable
~ $

**Comment by wileyhy** [comment]
"You'd have to try out all possible scenarios."

And isn't that why people say 'limit your scripts to 100 lines?' How else
could anyone find all the bugs? But then of course, if the legends are to
be believed, some corporations still maintain 20,000 line shell programs on
(pick your favorite obscure Unix flavor here). Bourne shell really wasn't
written with error handling in mind. I wish it was. You could rewrite it in
csh, though, right? ;-)

**Comment by wileyhy** [comment]
I'm sorry. There might be a solution. Someone else more knowledgeable
than me should take it from here.
  Wiley
